src/TextFieldCatalog.js

```diff
@@ -7,56 +7,84 @@ import ReactGA from 'react-ga';
         {leading && <i className='material-icons mdc-text-field__icon'>event</i>}
-        {textarea ? <TextArea textFieldId={textFieldId} onClick={onClick} /> : <Input textFieldId={textFieldId} onClick={onClick}/>}
-        {outlined || textarea ? null : <Label textFieldId={textFieldId} dense={dense}/>}
+        {textarea ?
+          <TextArea textFieldId={textFieldId} onClick={onClick} maxLength={maxLengthValue} placeholder={placeholder} noLabel={noLabel} /> :
+          <Input textFieldId={textFieldId} onClick={onClick} maxLength={maxLengthValue} placeholder={placeholder} noLabel={noLabel} />}
+        {outlined || textarea || noLabel ? null : <Label textFieldId={textFieldId} dense={dense}/>}
         {trailing && <i className='material-icons mdc-text-field__icon'>delete</i>}
-        {outlined || textarea ? <Outline textFieldId={textFieldId}/> : <div className='mdc-line-ripple'></div>}
+        {outlined || textarea ? <Outline noLabel={noLabel} textFieldId={textFieldId}/> : <div className='mdc-line-ripple'></div>}
       </div>
-      {helperText ? <HelperText /> : null}
+      { hasHelperLine ?
+        <div className='mdc-text-field-helper-line'>
+          {helperText ? <HelperText /> : null}
+          {characterCounter && !textarea ? <CharacterCounter /> : null}
+        </div>
+        : null
+      }
     </div>
   );
 }
 
-const FullWidthTextField = ({textarea, textFieldId, helperText}) => {
-  const classes = classnames('mdc-text-field', 'text-field', 'mdc-text-field--fullwidth', {
+const FullWidthTextField = ({textarea, textFieldId, helperText, characterCounter}) => {
+  // TODO: Remove --no-label class name in next release.
+  const classes = classnames('mdc-text-field', 'text-field', 'mdc-text-field--fullwidth', 'mdc-text-field--no-label', {
     'mdc-text-field--textarea': textarea,
   });
+  const hasHelperLine = helperText || characterCounter;
+
   return (
     <div className='text-field-container'>
       <div className={classes} ref={textFieldEl => textFieldEl && new MDCTextField(textFieldEl)}>
+        {characterCounter && textarea ? <CharacterCounter /> : null}
         {textarea ?
           <TextArea textFieldId={textFieldId}/> :
           <Input placeholder='Standard' textFieldId={textFieldId}/>}
         {textarea ? <Outline textFieldId={textFieldId}/> : null}
       </div>
-      {helperText ? <HelperText /> : null}
+      { hasHelperLine ?
+        <div className='mdc-text-field-helper-line'>
+          {helperText ? <HelperText /> : null}
+          {characterCounter && !textarea ? <CharacterCounter /> : null}
+        </div>
+        : null
+      }
     </div>
   );
 }
 
-const Outline = ({textFieldId}) => (
+const Outline = ({textFieldId, noLabel}) => (
   <div className='mdc-notched-outline' key='outline'>
     <div className='mdc-notched-outline__leading'></div>
-    <div className='mdc-notched-outline__notch'>
-      <Label textFieldId={textFieldId}/>
-    </div>
+    {noLabel ?
```

Comment:
Shouldn't we still be rendering the notch element, just not the label inside?

Reply:
Notch element isn't necessary. See outlined example [without label](https://github.com/material-components/material-components-web/tree/master/packages/mdc-textfield#outlined-1) here.